code/functions.py

```diff
@@ -0,0 +1,165 @@
+szsz = np.kron(sigmaz, sigmaz)
+
+
+def spectrum(sys, xticks, yticks, xdim, ydim, xlims=None, ylims=None, **kwargs):
+    p = SimpleNamespace(**kwargs)
+    
+    for key, value in kwargs.items():
+        try:
+            if len(value) > 0:
+                changing_variable = key
+                changing_values = value
+        except:
+            TypeError
+    
+    def energy(x):
+        p.__dict__[changing_variable] = x
+        H = sys.hamiltonian_submatrix(args=[p])
+        return np.linalg.eigvalsh(H)
+    
+    spectrum = np.array([energy(x) for x in changing_values])
+    
+    plot = hv.Path((changing_values, spectrum), kdims=[xdim, ydim])
+    
+    if xlims is None:
+        xlims = slice(xticks[0], xticks[-1])
+    else:
+        xlims = slice(xlims[0], xlims[1])
+    if ylims is None:
+        ylims = slice(yticks[0], yticks[-1], None)
+    else:
+        ylims = slice(ylims[0], ylims[1])
+    
+    return plot[xlims, ylims](plot={'xticks':xticks, 'yticks':yticks})
+
+def h_k(sys, p, phase_fac=1):
+    h, t = sys.cell_hamiltonian(args=[p]), sys.inter_cell_hopping(args=[p])
+    return h + t * phase_fac + t.T.conj() * np.conjugate(phase_fac)
+
+def find_pfaffian(H):
+    return np.sign(np.real(pf.pfaffian(1j*H)))
+
+def plot_bands(lead, p, momenta=np.linspace(-np.pi, np.pi)):
+    bands = kwant.physics.Bands(lead, args=[p])
+    evs = np.array([bands(k=k) for k in momenta])
+    return hv.Path((momenta, evs), kdims=[r'$k$', r'$E$'])
+
+def kitaev_chain(L=None, periodic=False):
+    lat = kwant.lattice.chain()
+    
+    if L is None:
+        sys=kwant.Builder(kwant.TranslationalSymmetry((-1,)))
+        L = 1
+    else:
+        sys = kwant.Builder()
+
+    # transformation to antisymmetric basis
+    U = np.matrix([[1.0, 1.0], [1.j, -1.j]]) / np.sqrt(2)
+    
+    onsite = lambda onsite, p: - p.mu * U.dot(sigmaz.dot(U.H))
+    for x in range(L):
+        sys[lat(x)] = onsite
+         
+    hop = lambda site1, site2, p: U.dot((-p.t * sigmaz - 1j * p.delta * sigmay).dot(U.H))
+    sys[kwant.HoppingKind((1,), lat)] = hop
+    
+    if periodic:
+        last_hop = lambda site1, site2, p: hop(site1, site2, p) * (1 - 2 * p.lambda_)
+        sys[lat(0), lat(L-1)] = last_hop
+    return sys
+
+def spinful_kitaev_chain(L=None, periodic=False):
+    lat = kwant.lattice.chain()
+    
+    if L is None:
+        sys=kwant.Builder(kwant.TranslationalSymmetry((-1,)))
+        L = 1
+    else:
+        sys = kwant.Builder()
+
+    # transformation to antisymmetric basis
+    onsite = lambda onsite, p: (2*p.t - p.mu) * szs0 + p.B * szsz
+    for x in range(L):
+        sys[lat(x)] = onsite
+         
+    hop = lambda site1, site2, p: -p.t * szs0 - 1j * p.delta * sys0
+    sys[kwant.HoppingKind((1,), lat)] = hop
+    
+    if periodic:
+        last_hop = lambda site1, site2, p: hop(site1, site2, p) * (1 - 2 * p.lambda_)
+        sys[lat(0), lat(L-1)] = last_hop
+    return sys
+
+def nanowire_chain(L=None, periodic=False):
```

Comment:
I suggest a general rule, where students who want to do the homework only ever need to modify the corresponding notebook, and never need to dig into files. This, unfortunately, means that system creation cannot be factored out, unless it's some generic system.